### PR TITLE
update scripts

### PR DESCRIPTION
**Pull request by huyle0110**

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -22,28 +22,38 @@ var utils = {
...
-		return _.mapObject(keyValues, function(val, key) {
+	toAttributeUpdates: function (keyValues) {
+		return _.mapObject(keyValues, function (val, key) {
 			return {
 				Action: 'PUT',
 				Value: val
 			}
 		});
+	},
+    isEmpty: function (obj) {
```
Comment (vanminh0910) [inline lib/utils.js]: Don't need this function. Please use the underscore module to check. http://underscorejs.org/#isEmpty

var _ = require('underscore');
_.isEmpty()

File: scripts/create.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.create = (event, context, callback) => {
+  const timestamp = new Date().getTime();
+  var input = JSON.parse(event.body);
+  const userId = event.requestContext.authorizer.principalId;
+  
+  if (!input.name || !input.xml || !input.lua ) {
```
Comment (vanminh0910) [inline scripts/create.js]: We allow script with empty xml or lua

File: scripts/create.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    callback(null, utils.createResponse(400, 'Please enter valid name , xml , lua of script'));
```
Comment (vanminh0910) [inline scripts/create.js]: When name is empty, use message 'Invalid script data'

File: scripts/create.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    return;
+  }
+  var params = {
```
Comment (vanminh0910) [inline scripts/create.js]: Please add blank line above this

File: scripts/create.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      lua: input.lua,
+      ownerId: userId,
+      createdAt: timestamp,
+      updatedAt: timestamp,
+    },
+  };
+
+  // write the script to the database
+  dynamodb.put(params, (error) => {
+    // handle potential errors
```
Comment (vanminh0910) [inline scripts/create.js]: Remove this comment

File: scripts/create.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    if (error) {
+      console.error(error);
+      callback(null, utils.createResponse(500, 'An internal server error occurred'));
+      return;
+    }
+
+    // create a response
+    callback(null, utils.createResponse(200,null, {
```
Comment (vanminh0910) [inline scripts/create.js]: Only return the script itself like this:

callback(null, utils.createResponse(200,null, params.Item));

File: scripts/delete.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+const jwt = require('jsonwebtoken');
+const config = require('../lib/config');
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.delete = (event, context, callback) => {
+    console.log(event);
+  const timestamp = new Date().getTime();
+  const email = event.requestContext.authorizer.principalId;
```
Comment (vanminh0910) [inline scripts/delete.js]: this is userId, not email

File: scripts/delete.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+const validator = require('validator');
+const bcrypt = require('bcrypt-nodejs');
+const jwt = require('jsonwebtoken');
+const config = require('../lib/config');
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.delete = (event, context, callback) => {
+    console.log(event);
```
Comment (vanminh0910) [inline scripts/delete.js]: please correct indentation

File: scripts/delete.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      ownerId: email,
+    },
+  };
+  dynamodb.get(getParams, (error, result) => {
+    if (error) {
+      console.error(error);
+      callback(null, utils.createResponse(500, 'Internal Server Error.'));
+      return;
+    }
+    if (utils.isEmpty(result)) {
```
Comment (vanminh0910) [inline scripts/delete.js]: Add blank line between clode blocks

File: scripts/delete.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      callback(null, utils.createResponse(400, 'Non-existing script'));
```
Comment (vanminh0910) [inline scripts/delete.js]: Please change message to be 'Script not exist'

File: scripts/delete.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        ownerId: email,
+      },
+      ConditionExpression: "id = :idDeleted",
+      ExpressionAttributeValues: {
+        ":idDeleted": event.pathParameters.id
+      },
+    };
+    // delete the script from the database
+    dynamodb.delete(params, (error, result) => {
+      // handle potential errors
```
Comment (vanminh0910) [inline scripts/delete.js]: Remove this comment, it's useless

File: scripts/delete.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      if (error) {
+         callback(null, utils.createResponse(500, 'An internal server error occurred'));
+        return;
+      }
+      // create a response
+      callback(null, utils.createResponse(500, 'Deleted successfully'));
```
Comment (vanminh0910) [inline scripts/delete.js]: If the request is successful, why do you return code 500?

File: scripts/list.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.list = (event, context, callback) => {
+  const ownerId = event.requestContext.authorizer.principalId;
+  var params = {
+    TableName: process.env.SCRIPTS_TABLE_NAME,
+    KeyConditionExpression: "#ownersub = :a",
+    ExpressionAttributeNames:{
+      "#ownersub": "ownerId"
```
Comment (vanminh0910) [inline scripts/list.js]: Should change to "#ownerId": "ownerId"

File: scripts/list.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+const config = require('../lib/config');
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.list = (event, context, callback) => {
+  const ownerId = event.requestContext.authorizer.principalId;
+  var params = {
+    TableName: process.env.SCRIPTS_TABLE_NAME,
+    KeyConditionExpression: "#ownersub = :a",
```
Comment (vanminh0910) [inline scripts/list.js]: a is not a good parameter name

File: scripts/list.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+const shortid = require('shortid');
+const validator = require('validator');
+const bcrypt = require('bcrypt-nodejs');
+const jwt = require('jsonwebtoken');
+const config = require('../lib/config');
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.list = (event, context, callback) => {
```
Comment (vanminh0910) [inline scripts/list.js]: Should rename list to listByUser

File: scripts/list.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  // list scripts from the database by userId
+  dynamodb.query(params, (error, result) => {
+    // handle potential errors
+    if (error) {
+      console.error(error);
+      callback(null, utils.createResponse(500, 'Please enter valid new name , new lua , new XML'));   
+      return;
+    }
+    // create a response
+    callback(null, utils.createResponse(200,null,result.Items));
```
Comment (vanminh0910) [inline scripts/list.js]: Missing spaces, should be (200, null, result.Items)

File: scripts/update.js [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+const config = require('../lib/config');
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.update = (event, context, callback) => {
+  const timestamp = new Date().getTime();
+  const input = JSON.parse(event.body);
+  const userId = event.requestContext.authorizer.principalId;
+  if (!input.newName || !input.newLua ||!input.newXml) {
```
Comment (vanminh0910) [inline scripts/update.js]: Allow blank script

File: scripts/update.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  const getParams = {
+    TableName: process.env.SCRIPTS_TABLE_NAME,
+    Key: {
+      id: event.pathParameters.id,
+      ownerId: userId,
+    }
+  };
+  dynamodb.get(getParams, (error, result) => {
```
Comment (vanminh0910) [inline scripts/update.js]: please add blank line between code blocks

File: scripts/update.js [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+const constants = require('../lib/constants');
+const dynamodb = require('../lib/dynamodb');
+const utils = require('../lib/utils');
+
+module.exports.update = (event, context, callback) => {
+  const timestamp = new Date().getTime();
+  const input = JSON.parse(event.body);
+  const userId = event.requestContext.authorizer.principalId;
+  if (!input.newName || !input.newLua ||!input.newXml) {
+    callback(null, utils.createResponse(400, 'Please enter valid new name , new lua , new XML'));
```
Comment (vanminh0910) [inline scripts/update.js]: Change message to be Invalid script data

File: scripts/update.js [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    TableName: process.env.SCRIPTS_TABLE_NAME,
+    Key: {
+      id: event.pathParameters.id,
+      ownerId: userId,
+    }
+  };
+  dynamodb.get(getParams, (error, result) => {
+    if (error) {
+      console.error(error);
+      callback(null, utils.createResponse(500, 'Internal Server Error.'));
```
Comment (vanminh0910) [inline scripts/update.js]: This message is not same as other 500 error message

File: scripts/update.js [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+      return;
+    }
+    if (utils.isEmpty(result)) {
```
Comment (vanminh0910) [inline scripts/update.js]: Use isEmpty() from underscore module

File: scripts/update.js [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+      callback(null, utils.createResponse(400, 'Non-existing script'));
+        return;
+    }
+    const updateParams = {
```
Comment (vanminh0910) [inline scripts/update.js]: Please follow my code in users for updating item.

File: scripts/update.js [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+      },
+      UpdateExpression: 'SET #name = :newName, #xml = :newXml,#lua =:newLua , updatedAt = :updatedAt',
+      ReturnValues: 'ALL_NEW',
+    };
+    // update the todo in the database
+    dynamodb.update(updateParams, (error, result) => {
+      // handle potential errors
+      if (error) {
+        console.error(error);
+        callback(new Error('Couldn\'t update the this item.'));
```
Comment (vanminh0910) [inline scripts/update.js]: You are not following my code

File: scripts/update.js [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+        return;
+      }
+      // create a response
+       callback(null, utils.createResponse(200,null,result ));
```
Comment (vanminh0910) [inline scripts/update.js]: You are not following my code

File: lib/utils.js (hunk 1 above)
Comment (huyle0110) [reply lib/utils.js]: fixed

File: scripts/create.js (hunk 2 above)
Comment (huyle0110) [reply scripts/create.js]: fixed

File: scripts/create.js (hunk 3 above)
Comment (huyle0110) [reply scripts/create.js]: fixed

File: scripts/create.js (hunk 4 above)
Comment (huyle0110) [reply scripts/create.js]: fixed

File: scripts/create.js (hunk 5 above)
Comment (huyle0110) [reply scripts/create.js]: fixed

File: scripts/create.js (hunk 6 above)
Comment (huyle0110) [reply scripts/create.js]: fixed

File: scripts/delete.js (hunk 7 above)
Comment (huyle0110) [reply scripts/delete.js]: fix

File: scripts/delete.js (hunk 8 above)
Comment (huyle0110) [reply scripts/delete.js]: fixed

File: scripts/delete.js (hunk 9 above)
Comment (huyle0110) [reply scripts/delete.js]: fixed

File: scripts/delete.js (hunk 10 above)
Comment (huyle0110) [reply scripts/delete.js]: fixed

File: scripts/delete.js (hunk 11 above)
Comment (huyle0110) [reply scripts/delete.js]: fixed

File: scripts/delete.js (hunk 12 above)
Comment (huyle0110) [reply scripts/delete.js]: fixed

File: scripts/update.js (hunk 21 above)
Comment (huyle0110) [reply scripts/update.js]: fixed

File: scripts/update.js (hunk 19 above)
Comment (huyle0110) [reply scripts/update.js]: fixed

File: scripts/create.js [hunk 25]
```diff
@@ -32,18 +33,14 @@ module.exports.create = (event, context, callback) => {
...
       console.error(error);
       callback(null, utils.createResponse(500, 'An internal server error occurred'));
       return;
     }
 
-    // create a response
-    callback(null, utils.createResponse(200,null, {
-      script : params.Item
-    }));
+    callback(null, utils.createResponse(200,null, params.Item
```
Comment (vanminh0910) [inline scripts/create.js]: No need to wrap the line here

File: scripts/delete.js [hunk 26]
```diff
@@ -9,49 +9,51 @@ const config = require('../lib/config');
...
       console.error(error);
       callback(null, utils.createResponse(500, 'Internal Server Error.'));
       return;
     }
-    if (utils.isEmpty(result)) {
-      callback(null, utils.createResponse(400, 'Non-existing script'));
-        return;
+
+    if (_.isEmpty(result)) {
+      callback(null, utils.createResponse(400, 'script not-exist'));
```
Comment (vanminh0910) [inline scripts/delete.js]: Message should be Script not exist

File: scripts/delete.js [hunk 27]
```diff
@@ -9,49 +9,51 @@ const config = require('../lib/config');
...
-      // handle potential errors
       if (error) {
-         callback(null, utils.createResponse(500, 'An internal server error occurred'));
+        callback(null, utils.createResponse(500, 'An internal server error occurred'));
         return;
       }
-      // create a response
-      callback(null, utils.createResponse(500, 'Deleted successfully'));
+
+      callback(null, utils.createResponse(200, 'Deleted successfully'));
```
Comment (vanminh0910) [inline scripts/delete.js]: If return 200, no need message

File: scripts/list.js [hunk 28]
```diff
@@ -10,27 +10,26 @@ const constants = require('../lib/constants');
 const dynamodb = require('../lib/dynamodb');
 const utils = require('../lib/utils');
 
-module.exports.list = (event, context, callback) => {
+module.exports.listbyuser = (event, context, callback) => {
```
Comment (vanminh0910) [inline scripts/list.js]: listByUser

File: scripts/list.js [hunk 29]
```diff
@@ -10,27 +10,26 @@ const constants = require('../lib/constants');
...
-    // handle potential errors
     if (error) {
       console.error(error);
-      callback(null, utils.createResponse(500, 'Please enter valid new name , new lua , new XML'));   
+      callback(null, utils.createResponse(500, 'An internal server error occurred '));   
       return;
     }
-    // create a response
-    callback(null, utils.createResponse(200,null,result.Items));
+    callback(null, utils.createResponse(200,null, result.Items));
```
Comment (vanminh0910) [inline scripts/list.js]: Add space between (200, null, result.Items)

File: scripts/update.js [hunk 30]
```diff
@@ -26,46 +28,47 @@ module.exports.update = (event, context, callback) => {
...
-      callback(null, utils.createResponse(500, 'Internal Server Error.'));
+      callback(null, utils.createResponse(500, 'An internal server error occurred '));
       return;
     }
-    if (utils.isEmpty(result)) {
-      callback(null, utils.createResponse(400, 'Non-existing script'));
-        return;
+
+    if (_.isEmpty(result)) {
+      callback(null, utils.createResponse(400, 'script not-exist'));
```
Comment (vanminh0910) [inline scripts/update.js]: Script not exist

File: scripts/update.js [hunk 31]
```diff
@@ -26,46 +28,47 @@ module.exports.update = (event, context, callback) => {
...
+    },
+    ReturnValues: 'ALL_NEW'
+  }, (error, result) => {
+    if (error) {
+      console.error(error);
+      callback(null, utils.createResponse(500, 'An internal server error occurred'));
+      return;
+    }
+
+    result.Attributes.password = '';
```
Comment (vanminh0910) [inline scripts/update.js]: This is wrongly copied from users

**Review by vanminh0910 — APPROVED** [review]

**Pull request merged**